7739 Dobbeltregistrering av helseperiode

ragnarwestad opened the pull request:
Fixen hindrer dobbeltregistrering av helsedekningperiode ved dobbeltklikk på "Bekreft og fortsett" i "Inngang" steget.

https://jira.adeo.no/browse/MELOSYS-7739

Bruker redux PENDING-status:

  1. reducers.ts - La til PENDING-håndtering
  2. selectors.ts - La til erPending selector

La til test for å verifiserer at ved dobbeltklikk kalles dispatch kun én gang (ikke to).

ragnarwestad committed:
7739 Hindre dobbeltregistrering av helsedekningperiode ved dobbeltklikk på "Bekreft og fortsett" i "Inngang" steget.

  Bruker useRef for å blokkere multiple klikk før første API-kall er ferdig.

oyvinyg reviewed (approved):
LGTM
Kan også være lurt å legge på en loading på selve knappen: 
```
      <Mui.StegKnapper
        bekreftKnappProps={{
          disabled: !redigerbart || !formIsValid,
          loading: isSubmittingRef.current,
          onClick: bekreftOgFortsett,
        }}
      />
```

thanglyphan commented on src/sider/eu_eøs/pensjonist/stegKomponenter/vurderingOpplysninger/vurderingOpplysninger.test.tsx:
```diff
@@ -507,6 +507,44 @@ describe("VurderingOpplysninger - MELOSYS-7642", () => {
     });
   });
 
+  describe("MELOSYS-7739: Dobbeltklikk-beskyttelse", () => {
```
Trenger ikke ha jiranummer på testen, bare tittel på hva den tester holder

thanglyphan commented on src/sider/eu_eøs/pensjonist/stegKomponenter/vurderingOpplysninger/vurderingOpplysninger.tsx:
```diff
@@ -124,7 +125,10 @@ export function VurderingOpplysninger({ bekreft, oppdaterStatus, aktivtSteg }: P
 
   const bekreftOgFortsett = async () => {
+    if (isSubmittingRef.current) return;
+    isSubmittingRef.current = true;
```
Knappen kan være disablet i det man trykker for første gang, altså når dispatch fyres av, trenger vi denne ref håndteringen? 

Når den hentes, skal denne være pending, ref: 

```
interface PendingAction {
  type: typeof PENDING;
}

```

Er det et alternativ å bruke denne for å toggle disabled i knappen?

ragnarwestad replied:
Ja, god ide som vanlig Thang. Jeg endret det til å bruke PENDING, det er jo en mye bedre løsning.

thanglyphan reviewed (changes requested):
Har noen kommentarer, avklar før merge

ragnarwestad committed:
Endret fra useRef-basert dobbeltklikk-beskyttelse til redux PENDING-status:

  1. reducers.ts - La til PENDING-håndtering

  case Types.PENDING:
    return { ...state, status: STATUS.PENDING };

  2. selectors.ts - La til isPending selector

  export const helseutgiftDekkesPeriodeIsPendingSelector = ...

  3. vurderingOpplysninger.tsx - Erstattet useRef med redux

  - Fjernet useRef import og isSubmittingRef
  - La til isPending selector
  - Sjekker isPending i bekreftOgFortsett()
  - Knappen er nå disabled når isPending er true

  4. Tester oppdatert

  - Nye tester verifiserer at knappen er disabled når status er PENDING
  - Tester at dispatch ikke kalles når isPending er true

  Fordeler:
  - Renere kode - følger eksisterende redux-mønster
  - Knappen viser visuelt at den er disabled (bedre UX)
  - Konsistent med resten av kodebasen

ragnarwestad committed:
7739 Fjernet overflødig isPending-sjekk i bekreftOgFortsett

  Knappen er allerede disabled={... || isPending}, så funksjonen
  kan ikke kalles når isPending er true. Sjekken var unødvendig.

thanglyphan commented on src/sider/eu_eøs/pensjonist/stegKomponenter/vurderingOpplysninger/vurderingOpplysninger.tsx:
```diff
@@ -41,6 +41,7 @@ export function VurderingOpplysninger({ bekreft, oppdaterStatus, aktivtSteg }: P
   const [visOppfrisk, setVisOppfrisk] = useState(false);
 
   const behandlingID = useSelector(behandlingerSelectors.BehandlingIDSelector);
+  const isPending = useSelector(helseutgiftDekkesPeriodeSelector.HelseutgiftDekkesPeriodeErPendingSelector);
```
Gi isPending ett nytt og mer forklarende navn. Hold gjerne det til norsk, erPending isåfall. oppdatererHelseUtgiftDekkesPeriode eller noe i den duren.

ragnarwestad committed:
7739 Korrigert navn

ragnarwestad committed:
7739 Bruker status konstant i stedet for string

thanglyphan commented on src/sider/eu_eøs/pensjonist/stegKomponenter/vurderingOpplysninger/vurderingOpplysninger.tsx:
```diff
@@ -41,6 +41,7 @@ export function VurderingOpplysninger({ bekreft, oppdaterStatus, aktivtSteg }: P
   const [visOppfrisk, setVisOppfrisk] = useState(false);
 
   const behandlingID = useSelector(behandlingerSelectors.BehandlingIDSelector);
+  const erPending = useSelector(helseutgiftDekkesPeriodeSelector.HelseutgiftDekkesPeriodeErPendingSelector);
```
Denne beskriver ikke hva vi venter på. Vil helst at variablene våre skal være beskrivende nok for andre til å skjønne hva denne gjør. Sleng på hva den er pending på, f.eks: `helseutgiftDekkesPeriodeErPending`

ragnarwestad replied:
Fikset, men ikke foreslå alternativer du ikke aksepterer da, da kaster vi bare bort tid. 

Vennligst fjern blokkeringen nå så vi kan få denne inn.

ragnarwestad committed:
7739 Korrigert navn

thanglyphan commented:
Legg til jirasaken i beskrivelse før du merger
